design_drafts/controller_execution_management.md

```diff
@@ -0,0 +1,125 @@
+# Controller Execution Management
+
+The following shall discuss the execution of loaded controllers inside the `ControllerManager`.
+As of today, the loaded controllers are processed in a sequential order which doesn't allow for much flexibility or interoperability between the controllers.
+We identified a few essential criteria when it comes to the execution of controllers:
+
+* Sequential Execution of Controllers
+* Parallel Execution of Controllers
+* Nesting Controllers
+* Variable Frequencies of Controllers
+
+We'll go into more details below for each aspect.
+
+
+## Prerequisites
+
+### Flexible Joint State Message
+
+Over the course of this document, we are referring to a "Flexible Joint State Message".
+This data type is precisely described in [here](flexible_joint_states_msg.md).
+
+What is important to understand for this article is that this joint state message serves as a general container structure for various, dynamically specified key values pairs.
+Every controller can hereby access values which are specific to itself without being limited to a set of pre-defined keys.
+For example can a `JointEffortController` access the robot's input state via keys like `joint_position` and `joint_velocity` and write its output torque under a key such as `torque_command`.
+This further allows an easy extensibility in the variety of controllers which could access data completely independent of their underlying hardware.
+Various components such as camera systems or third-party peripherals can therefore be combined and shared between controllers.
+A typical example for this are visuomotor controllers where camera images are used to generate torque values.
+Relevant features extracted from camera inputs can so be shared between multiple controllers to eventually calculate the appropriate control command.
+
+### Controller and Hardware Interface
+
+Each controller gets the previously mentioned joint state message attached during their initialization phase.
+The controller manager is responsible to initialize the joint state message appropriately and route it correctly to each controller.
+It is further the task of the controller manager to fill the message with a call to `read` from the connected hardware interface and apply the results from the controllers accordingly with a call to `write`.
```

Comment:
Does it do this by passing the setpoint to each individual controller or should we account for a case where some calculation needs to be performed?

Reply:
Not sure I understand your comment. Can you please elaborate?

What I tried to say here is that the controller manager is sole entry point for communicating with the actual underlying hardware. The robot's state can be fetch by a call to `read() -> flexible_joint_message` which creates/fills the joint state message. It is then passed to the controllers, which perform an update and the final joint state message is then being passed to the hardware to apply their values. All calculations/modifications to the map shall be done in the controllers exclusively.

Reply:
Thanks for the clarification!
It seems my concern here is more applicable to my question on parallel controller container frequencies below and is unrelated to the responsibility of the controller manager.